app/src/main/java/org/dicekeys/app/encryption/BiometricsHelper.kt

```diff
@@ -0,0 +1,122 @@
+
+            // TODO change texts
+
+            val promptInfo = BiometricPrompt.PromptInfo.Builder()
+                    .setTitle("Biometric Encryption")
+                    .setSubtitle("Log in using your biometric credential")
+                    .setDescription("This is a description for the action")
+                    .setNegativeButtonText(fragment.getString(android.R.string.cancel))
+                    .setConfirmationRequired(true)
+                    .setAllowedAuthenticators(BiometricManager.Authenticators.BIOMETRIC_STRONG)
+                    .build()
+
+            val biometricPrompt = BiometricPrompt(fragment, ContextCompat.getMainExecutor(fragment.requireContext()),
+                    object : BiometricPrompt.AuthenticationCallback() {
+                        override fun onAuthenticationError(
+                                errorCode: Int,
+                                errString: CharSequence
+                        ) {
+                            super.onAuthenticationError(errorCode, errString)
+
+                            if (errorCode == BiometricPrompt.ERROR_USER_CANCELED || errorCode == BiometricPrompt.ERROR_NEGATIVE_BUTTON || errorCode == BiometricPrompt.ERROR_CANCELED) {
+                                // This is OK
+                            } else {
+                                fragment.toast("Authentication error:$errorCode $errString")
+                            }
+                        }
+
+                        override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
+                            super.onAuthenticationSucceeded(result)
+
+                            result.cryptoObject?.cipher?.let { cipher ->
+                                // Encrypt DiceKey
+                                val encryptedData = appKeystore.encryptData(cipher, diceKey.toHumanReadableForm().toByteArray())
+                                // Save data
+                                encryptedStorage.save(diceKey, encryptedData)
+                            }
+                        }
+
+                        override fun onAuthenticationFailed() {
+                            super.onAuthenticationFailed()
+                            fragment.toast("Authentication failed")
+                        }
+                    })
+
+            try {
+                biometricPrompt.authenticate(
+                        promptInfo,
+                        BiometricPrompt.CryptoObject(appKeystore.getBiometricsEncryptionCipher())
+                )
+            } catch (e: Exception) {
+                e.printStackTrace()
+            }
+
+        } else {
+            fragment.toast("You have to enable Biometrics to use this feature.")
+        }
+    }
+
+    fun decrypt(encryptedDiceKey: EncryptedDiceKey, fragment: Fragment, success: (diceKey: DiceKey<*>)->Unit) {
+        // TODO change texts
+        val promptInfo = BiometricPrompt.PromptInfo.Builder()
```

Comment:
Does this need to be replicated in two functions or can it be common?

Reply:
We can make it a bit more compact, but there are different texts and options when decrypting / encrypting.